File: MvvmSample/app/src/main/java/com/abben/mvvmsample/util/RateLimiter.java
package com.abben.mvvmsample.util;

import android.os.SystemClock;
import android.support.v4.util.ArrayMap;

import java.util.concurrent.TimeUnit;

/**
 * 决定是否应该获取一些数据的Utility类
 * 根据现在离上一次获取数据的时间来决定
 * Created by abben on 2017/10/25.
 */
public class RateLimiter<KEY> {
    private ArrayMap<KEY, Long> timestamps = new ArrayMap<>();
    private final long timeout;

    public RateLimiter(int timeout, TimeUnit timeUnit) {
        this.timeout = timeUnit.toMillis(timeout);
    }

    public synchronized boolean shouldFetch(KEY key) {
        Long lastFetched = timestamps.get(key);
        long now = now();
        if (lastFetched == null) {
            timestamps.put(key, now);
            return true;
        }
        if (now - lastFetched > timeout) {
            timestamps.put(key, now);
            return true;
        }
        return false;
    }

    private long now() {
        return SystemClock.uptimeMillis();
    }

    public synchronized void reset(KEY key) {
        timestamps.remove(key);
    }
}

File: MvvmSampleJetpack/app/src/main/java/com/abben/mvvmsamplejetpack/util/RateLimiter.kt
package com.abben.mvvmsamplejetpack.util

import android.os.SystemClock
import androidx.collection.ArrayMap
import java.util.concurrent.TimeUnit

/**
 * 决定是否应该获取一些数据的Utility类
 * 根据现在离上一次获取数据的时间来决定
 *  @author abben
 *  on 2019/1/18
 */
class RateLimiter<KEY>(timeout: Int, timeUnit: TimeUnit) {
    private val timestamps: ArrayMap<KEY, Long> = ArrayMap()
    private val timeout = timeUnit.toMillis(timeout.toLong())

    @Synchronized
    fun shouldFetch(key: KEY): Boolean {
        val lastFetched = timestamps[key]
        val now = now()
        return if (lastFetched == null) {
            timestamps[key] = now
            true
        } else {
            if (now - lastFetched > timeout) {
                timestamps[key] = now
                true
            } else {
                false
            }
        }
    }

    private fun now(): Long {
        return SystemClock.uptimeMillis()
    }

    fun reset(key: KEY) {
        timestamps.remove(key)
    }

}